### PR TITLE
[WIP] Bugfix/cflags requires

### LINKED ISSUE #39671: Compiler flags in spec requirements regression
### Steps to reproduce

Create the following environment:

```
spack:
  specs:
  - zlib
  view: false
  concretizer:
    unify: true
  packages:
    zlib:
      require: cflags="-g"
```

with this environment active, `spack concretize` fails.

This is on d9d1eb24f9e1ef3550256ca16d5b0c567c3102c7

I can also reproduce it on 9bb5cff but not 87d4bdaa025b624d3b891f4e2f0c4034b502496b. So I'm guessing this logic needs to be updated with a better understanding of https://github.com/spack/spack/pull/38447.

What I find strange is that we have a test for this: `test_require_cflags`, which uses nearly exactly this config and passes (on all mentioned commits).

### Error message

```
$ spack -d concretize -f
...
==> [2023-08-28-18:03:55.024834] UnsatisfiableSpecError: concretization failed for the following reasons:

   1. cannot satisfy a requirement for package 'zlib'.
==> [2023-08-28-18:03:55.024946] Error: concretization failed for the following reasons:

   1. cannot satisfy a requirement for package 'zlib'.
Traceback (most recent call last):
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/main.py", line 1023, in main
    return _main(argv)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/main.py", line 978, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/main.py", line 1006, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/main.py", line 647, in _invoke_command
    return_val = command(parser, args)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/cmd/concretize.py", line 44, in concretize
    concretized_specs = env.concretize(force=args.force, tests=tests)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/environment/environment.py", line 1333, in concretize
    return self._concretize_together(tests=tests)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/environment/environment.py", line 1428, in _concretize_together
    concrete_specs: List[spack.spec.Spec] = spack.concretize.concretize_specs_together(
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/concretize.py", line 741, in concretize_specs_together
    return _concretize_specs_together_new(*abstract_specs, **kwargs)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/concretize.py", line 748, in _concretize_specs_together_new
    result = solver.solve(abstract_specs, tests=kwargs.get("tests", False))
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/solver/asp.py", line 2944, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/solver/asp.py", line 864, in solve
    error_handler.raise_if_errors()
  File "/g/g17/scheibel/repos/spack-requireflags/spack/lib/spack/spack/solver/asp.py", line 683, in raise_if_errors
    raise UnsatisfiableSpecError(msg)
spack.solver.asp.UnsatisfiableSpecError: concretization failed for the following reasons:

   1. cannot satisfy a requirement for package 'zlib'.
```

### Information on your system

I can reproduce this on Linux and Mac, e.g.

```
$ spack debug report
* **Spack:** 0.21.0.dev0 (d9d1eb24f9e1ef3550256ca16d5b0c567c3102c7)
* **Python:** 3.9.12
* **Platform:** linux-rhel8-broadwell
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes https://github.com/spack/spack/issues/39671

It appears an extra layer of indirection has been added in the form of `condition_effect` and `condition_trigger` since the last time I tried to understand the concretizer (between the requirement group member and the imposed constraint).

I observed in #39671 that our basic test was passing even though a simple example outside of the tests was failing, and I can say that is because:

* we use separate numbering spaces for the imposed constraints and requirement group members
* these spaces overlap (i.e. both start from 0)

I think we should use a global ID numbering system since I think all sorts of strange bugs could result from this.

### CODE REVIEW

File: lib/spack/spack/solver/concretize.lp [hunk 1]
```diff
@@ -689,15 +689,18 @@ requirement_group_satisfied(node(ID, Package), X) :-
...
+  activate_requirement(node(NodeID1, A1), X),
+  pkg_fact(A1, condition_effect(Y, Z)),
+  imposed_constraint(Z,"node_flag_set", A1, A2, A3).
+
+{ attr("node_flag_source", node(NodeID1, A1), A2, node(NodeID2, A3)) } :-
+  requirement_group_member(Y, A1, X),
+  activate_requirement(node(NodeID1, A1), X),
+  pkg_fact(A1, condition_effect(Y, Z)),
+  imposed_constraint(Z,"node_flag_source", A1, A2, A3),
+  imposed_packages(NodeID2, A3).
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp]: Thanks. There's clearly a bug that got unnoticed. I think this:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 9d1b48b379..6249503751 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -689,15 +689,17 @@ requirement_group_satisfied(node(ID, Package), X) :-
 % flags if their only source is from a requirement. This is overly-specific
 % and should use a more-generic approach like in https://github.com/spack/spack/pull/37180
 
-{ attr("node_flag", node(ID, A1), A2, A3) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_set", A1, A2, A3).
+ { attr("node_flag", node(ID, A1), A2, A3) } :-
+  requirement_group_member(ConditionID, Package, RequirementID),
+  activate_requirement(node(ID, Package), RequirementID),
+  pkg_fact(Package, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID,"node_flag_set", A1, A2, A3).
 
 { attr("node_flag_source", node(ID, A1), A2, node(ID, A3)) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_source", A1, A2, A3).
+  requirement_group_member(ConditionID, Package, RequirementID),
+  activate_requirement(node(ID, Package), RequirementID),
+  pkg_fact(Package, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID,"node_flag_source", A1, A2, A3).
 
 requirement_weight(node(ID, Package), Group, W) :-
   W = #min {
```
would be more correct for cases where a virtual environment is involved:
```yaml
packages:
  mpi:
    require: openmpi cflags="-g"
```
In that case we have:
```
%----------------------------------------------------------------------------
% Imposed requirements
%----------------------------------------------------------------------------
requirement_group("mpi",0).
requirement_policy("mpi",0,"one_of").

pkg_fact("mpi",condition(0)).
condition_reason(0,"None").
pkg_fact("mpi",condition_trigger(0,0)).
pkg_fact("mpi",condition_effect(0,0)).
requirement_group_member(0,"mpi",0).
requirement_has_weight(0,0).


%----------------------------------------------------------------------------
% Trigger conditions
%----------------------------------------------------------------------------
pkg_fact("mpi",trigger_id(0)).
pkg_fact("mpi",trigger_msg("mpi")).
condition_requirement(0,"virtual_node","mpi").


%----------------------------------------------------------------------------
% Imposed requirements
%----------------------------------------------------------------------------
pkg_fact("mpi",effect_id(0)).
pkg_fact("mpi",effect_msg("openmpi cflags=\"-g\"")).
imposed_constraint(0,"node","openmpi").
imposed_constraint(0,"node_flag_set","openmpi","cflags","-g").
imposed_constraint(0,"node_flag_source","openmpi","cflags","openmpi").
```
so `Package` and `A1` don't always coincide.

File: lib/spack/spack/solver/concretize.lp [hunk 2]
```diff
@@ -689,15 +689,18 @@ requirement_group_satisfied(node(ID, Package), X) :-
...
-{ attr("node_flag", node(ID, A1), A2, A3) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_set", A1, A2, A3).
-
-{ attr("node_flag_source", node(ID, A1), A2, node(ID, A3)) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_source", A1, A2, A3).
+attr("node_flag", node(NodeID1, A1), A2, A3) :-
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp]: Here  I would still leave a choice to clingo, since this might be part of a "one_of" set of requirements.

File: lib/spack/spack/solver/concretize.lp (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp]: That said, the `mpi` case works also with this patch - I'll try to understand why since the names of the virtual "mpi" and of the package "openmpi" don't match the pattern above :thinking:

**Review by alalazo — COMMENTED** [review]
A few comments. I think we need more unit tests for these edge cases, I'll try to come up with some.

File: lib/spack/spack/solver/concretize.lp (hunk 2 above)
Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp]: Thanks, I couldn't think of why `{}` was needed. Apparently we don't test that.

**Comment by scheibelp** [comment]
> I think we need more unit tests for these edge cases

I think that applies for https://github.com/spack/spack/pull/39880#discussion_r1328613756, but more fundamentally, while we use overlapping ID systems, all of our tests will be fragile. `test_require_cflags` *should* be all that is needed to confirm this behavior (at least the basic ability to add cflags with requirements).

File: lib/spack/spack/solver/concretize.lp (hunk 1 above)
Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp]: Do virtual nodes have the same ID as the provider? The ID's are forced to be the same in your example:

```
attr("node_flag", node(ID, A1), A2, A3) } :-
...
+  activate_requirement(node(ID, Package), RequirementID),
```

which I assume would mean that `Package = A1` (to be clear, they must also be the same in my example, it's just that I think that further modification is required beyond what you suggested).

Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp]: > Do virtual nodes have the same ID as the provider?

Nope, you're correct. I wish we could simplify this rule but apparently flags have their own peculiar semantic :thinking:

File: lib/spack/spack/solver/concretize.lp [hunk 3]
```diff
@@ -689,15 +689,29 @@ requirement_group_satisfied(node(ID, Package), X) :-
...
+  activate_requirement(node(ID, SubPackage), RequirementID),
+  pkg_fact(SubPackage, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID, "node_flag_set", SubPackage, FlagType, FlagValue).
+
+{ attr("node_flag", node(SubPackageID, SubPackage), FlagType, FlagValue) } :-
+  requirement_group_member(ConditionID, Package, RequirementID),
+  activate_requirement(node(PackageID, Package), RequirementID),
+  pkg_fact(Package, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID, "node_flag_set", Package, FlagType, FlagValue),
+  provider(node(SubPackageID, SubPackage), node(PackageID, Package)).
```
Comment (scheibelp) [inline lib/spack/spack/solver/concretize.lp]: @alalazo

Based on  https://github.com/spack/spack/pull/39880#discussion_r1328593642 I added this, although the updated test doesn't pass (i.e. this is not enforced).

I added the below constraint, since it seemed like the package wasn't being marked as a provider by the concretizer:

```
1 { attr("virtual_node", node(0..X-1, Virtual)) : max_dupes(Virtual, X) } :- 
  virtual_condition_holds(node(ProviderID, Provider), Virtual),
  attr("node", node(ProviderID, Provider)).
```

it turns out with that constraint, you can't do `spack spec openmpi`, which seems wrong to me (my statement is "if there is a node and it meets the condition to provide a virtual, then there should be a virtual node for that").

I'm inclined to do one of:

* Drop support for cflags on requirements associated with virtuals (only allow it for non-virtuals)
* or, use a generic approach like https://github.com/spack/spack/pull/37180

Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp]: > it turns out with that constraint, you can't do `spack spec openmpi`

I should further clarify that this also fails when adding just this clause to the `concretize.lp` on develop (i.e. including none of the other constraints here).

**Comment by scheibelp** [comment]
If I read your case more carefully:

```
packages:
  mpi:
    require: openmpi cflags="-g"
```

you are specifically talking about requirements from virtuals that mention specific implementations. Part of what I'm tangling with based on your request is supporting things like:

```
packages:
  mpi:
    require: cflags="-g"
```

(i.e. propagating a requirement on a virtual to all possible providers).

If you only care about the former and not the latter, there may be a way to address this at the `asp.py` level vs. in `concretize.lp`, although it might require breaking down your example into

```
packages:
  mpi:
    require: openmpi
  openmpi:
    require: cflags="-g"
```

i.e. requirements on virtuals should only mention details relevant to virtuals and not providers.

File: lib/spack/spack/solver/concretize.lp [hunk 4]
```diff
@@ -509,6 +509,12 @@ attr("virtual_node", VirtualNode) :- attr("virtual_root", VirtualNode).
 % then the provider is the root package.
 attr("root", PackageNode) :- attr("virtual_root", VirtualNode), provider(PackageNode, VirtualNode).
 
+attr("virtual_root", VirtualNode) :- attr("root", PackageNode), provider(PackageNode, VirtualNode).
```
Comment (scheibelp) [inline lib/spack/spack/solver/concretize.lp:519]: This isn't needed, but seems reasonable.

File: lib/spack/spack/solver/concretize.lp [hunk 5]
```diff
@@ -19,30 +19,30 @@
...
+error(100, "e18") :- attr("depends_on", ParentNode, _, _), not attr("node", ParentNode).
+error(100, "e19") :- attr("depends_on", _, ChildNode, _), not attr("node", ChildNode).
+error(100, "e20") :- attr("node_flag_source", ParentNode, _, _), not attr("node", ParentNode).
+error(100, "e21") :- attr("node_flag_source", _, _, ChildNode), not attr("node", ChildNode).
+
+error(100, "e1") :- attr("virtual_node", VirtualNode), not provider(_, VirtualNode).
+error(100, "e2") :- provider(_, VirtualNode), not attr("virtual_node", VirtualNode).
+error(100, "e3") :- provider(PackageNode, _), not attr("node", PackageNode).
+
+error(100, "e4") :- attr("root", node(ID, PackageNode)), ID > min_dupe_id.
```
Comment (scheibelp) [inline lib/spack/spack/solver/concretize.lp:45]: Introducing these error facts helped me figure out why 

```
packages:
  virtual-w:
    require: cflags="-O1"
```

doesn't work (i.e. requirements from the virtual that do not explicitly mention an implementation): virtuals can't propagate most constraints to nodes.

File: lib/spack/spack/solver/concretize.lp (hunk 3 above)
Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp]: I'll mark this as resolved since other updates have shed some light on why this wasn't working.

**Comment by scheibelp** [comment]
The [CI failure](https://gitlab.spack.io/spack/spack/-/jobs/8399678) appears legitimate in the sense that it is cause by this PR, but I'll note that the associated `spack.yaml` appears to be malformed (`share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml`):

```
  packages:
    mpi:
      require: "mpich"
  specs:
  - openmpi
```

I'm not sure what was happening before, but I think this file should be updated.

**Comment by alalazo** [comment]
@scheibelp Fwiw it also seemed a bug to me, but this is why we support that https://github.com/spack/spack/pull/38447#discussion_r1293007160

**Comment by scheibelp** [comment]
> Fwiw it also seemed a bug to me, but this is why we support that https://github.com/spack/spack/pull/38447#discussion_r1293007160

There seems to be some openness in that thread to changing the semantics. That is what I would support here. The most succinct way I can express my disagreement is to say that our modeling in `concretize.lp` is not prepared to handle this desire in a conceptually consistent manner, i.e. that the perfectly reasonable expression:

```
1 { attr("virtual_node", node(0..X-1, Virtual)) : max_dupes(Virtual, X) } :-
  virtual_condition_holds(node(ProviderID, Provider), Virtual),
  attr("node", node(ProviderID, Provider)).
```

is the cause of the CI failure (I have reproduced this in https://github.com/spack/spack/pull/40261): we are relying on obscure and hard-to-explain choices made by the concretizer to support this case (namely, the concretizer succeeds by avoiding the creation of a virtual node, not something I'd like to depend on). If we want to support it, we should support it in a way that makes sense.

File: lib/spack/spack/solver/concretize.lp [hunk 6]
```diff
@@ -689,15 +696,28 @@ requirement_group_satisfied(node(ID, Package), X) :-
...
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_set", A1, A2, A3).
-
-{ attr("node_flag_source", node(ID, A1), A2, node(ID, A3)) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_source", A1, A2, A3).
+% None of the following 3 are needed for `virtual-w:require: w cflags="-O1"`
+% (1) and (3) are needed for `y:require: cflags="-g"`
```
Comment (scheibelp) [inline lib/spack/spack/solver/concretize.lp:729]: Regarding this and your point

> That said, the mpi case works also with this patch - I'll try to understand why since the names of the virtual "mpi" and of the package "openmpi" don't match the pattern above 

I will note that `emit_facts_from_requirement_rules` has differing behavior for virtuals vs. providers:

```
For y:

pkg_fact("y",trigger_id(0)).
pkg_fact("y",trigger_msg("y cflags=\"-g\"")).
condition_requirement(0,"node","y").
condition_requirement(0,"node_flag","y","cflags","-g").
condition_requirement(0,"node_flag_source","y","cflags","y").

For w:

pkg_fact("virtual-w",trigger_id(0)).
pkg_fact("virtual-w",trigger_msg("virtual-w")).
condition_requirement(0,"virtual_node","virtual-w").
```

the rule is easily satisfied for virtuals, but for non-virtuals the "condition" is equal to the constraint, I think that's to support groups, but IMO it ought to be refactored into something sensible.

Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp:729]: Update: none of these are required if using:

```

{ impose(EffectID, node(ID, Package)) } :-
  pkg_fact(Package, condition_effect(Y, EffectID)),
  requirement_group_member(Y, Package, X),
  activate_requirement(node(ID, Package), X),
  requirement_group(Package, X).
```

(added above)

File: lib/spack/spack/solver/concretize.lp (hunk 4 above)
Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp:519]: Hmm, I need to think about that. `virtual_root` was used to signal that the root spec is a virtual. Not sure in which context this new rule is helpful :thinking:

File: lib/spack/spack/solver/counter.py [hunk 7]
```diff
@@ -34,6 +34,9 @@ def __init__(self, specs: List["spack.spec.Spec"], tests: bool) -> None:
 
         self._possible_dependencies: PossibleDependencies = set()
         self._possible_virtuals: Set[str] = set(x.name for x in specs if x.virtual)
+        for x in specs:
+            if not x.virtual:
+                self._possible_virtuals.update(str(i) for i in x.package_class.provided.keys())
```
Comment (alalazo) [inline lib/spack/spack/solver/counter.py:39]: Ok, this would be part of the changes in semantic mentioned above. Not sure why we would need it though :thinking:

File: lib/spack/spack/solver/concretize.lp [hunk 8]
```diff
@@ -516,6 +516,13 @@ attr("virtual_node", VirtualNode) :- attr("virtual_root", VirtualNode).
 % then the provider is the root package.
 attr("root", PackageNode) :- attr("virtual_root", VirtualNode), provider(PackageNode, VirtualNode).
 
+attr("virtual_root", VirtualNode) :- attr("root", PackageNode), provider(PackageNode, VirtualNode).
+
+% This is required for require_cflags to pass with `require: w cflags="-O1"`
+1 { attr("virtual_node", node(0..X-1, Virtual)) : max_dupes(Virtual, X) } :-
+  virtual_condition_holds(node(ProviderID, Provider), Virtual),
+  attr("node", node(ProviderID, Provider)).
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp:524]: This will not blend with the semantic we want to merge in #35322. If a condition to provide a virtual holds, it is not necessary ~for the node to provide the virtual~ for the virtual node to be present.

File: lib/spack/spack/solver/concretize.lp [hunk 9]
```diff
@@ -674,11 +681,33 @@ activate_requirement(node(ID, Package), X) :-
   requirement_conditional(Package, X, Y).
 
 requirement_group_satisfied(node(ID, Package), X) :-
-  1 { condition_holds(Y, node(ID, Package)) : requirement_group_member(Y, Package, X) } 1,
+  1 { requirement_condition_satisfied(Y, node(ID, Package)) : requirement_group_member(Y, Package, X) } 1,
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp:684]: Just a reminder that we need to make "any_of" symmetric too, in case.

File: lib/spack/spack/solver/concretize.lp [hunk 10]
```diff
@@ -674,11 +681,33 @@ activate_requirement(node(ID, Package), X) :-
...
+  requirement_group_member(Y, Package, X),
+  activate_requirement(node(ID, Package), X),
+  requirement_group(Package, X).
+
+{ impose(EffectID, node(ProviderID, Provider)) } :-
+  pkg_fact(Provider, condition_effect(Y, EffectID)),
+  provider(node(ProviderID, Provider), node(VirtualID, Virtual)),
+  requirement_group_member(Y, Virtual, X),
+  activate_requirement(node(VirtualID, Virtual), X),
+  requirement_group(Virtual, X).
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp:709]: I'll look at this better, but I don't immediately see why we can give a choice to impose something without having anything that says that the corresponding trigger condition is satisfied in the body of the rule. 

The only explanation I see is that rules in requirements are always of the kind:
```
%----------------------------------------------------------------------------
% Trigger conditions
%----------------------------------------------------------------------------
pkg_fact("openmpi",trigger_id(0)).
pkg_fact("openmpi",trigger_msg("openmpi%gcc+static")).
condition_requirement(0,"node","openmpi").
condition_requirement(0,"variant_value","openmpi","static","True").
condition_requirement(0,"node_compiler","openmpi","gcc").
condition_requirement(0,"node_compiler_version_satisfies","openmpi","gcc",":").

%----------------------------------------------------------------------------
% Imposed requirements
%----------------------------------------------------------------------------
pkg_fact("openmpi",effect_id(0)).
pkg_fact("openmpi",effect_msg("openmpi%gcc+static")).
imposed_constraint(0,"node","openmpi").
imposed_constraint(0,"variant_set","openmpi","static","True").
imposed_constraint(0,"node_compiler_set","openmpi","gcc").
imposed_constraint(0,"node_compiler_version_satisfies","openmpi","gcc",":").
```
but that also means they work logically backwards with this choice - we give choice to impose the rules because they are designed in such a way that imposing the effect implies the causes.

File: lib/spack/spack/solver/concretize.lp (hunk 8 above)
Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp:524]: Two points:

* This only says that if X provides a virtual Y, that Y must exist as a virtual. It does not say that X must be the provider for Y. Therefore I don't think it actually conflicts (at least with the high-level goal stated in #35322)
* Without this, note that  `packages:<virtual>:require` cannot guarantee imposing requirements on providers

```
packages:
  mpi:
    require: openmpi +shared
```

i.e. this can only place constraints on MPI right now when MPI is a (virtual) node in the DAG. Another option is to take constraints like `require: openmpi+shared` and generate a secondary requirement for the provider (whether or not it is providing something).

File: lib/spack/spack/solver/concretize.lp (hunk 10 above)
Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp:709]: > I don't immediately see why we can give a choice to impose something without having anything that says that the corresponding trigger condition is satisfied in the body of the rule.

In this case there is no "valid" corresponding trigger condition:

* the condition requirements generated (the ones you list in the example directly above) are not actually for triggering the impose, but rather to record whether requirement group members are satisfied (technically they also trigger the impose, but they are effectively useless for that purpose)
* currently the only "influence" that the requirement group creates are error conditions when their members are not satisfied
* "somehow", we have a means to thwart error facts by deriving attributes for all requirements other than cflags
  * IMO, it is not worth understanding how or why this works, and saying that we are allowed to impose things if anyone might want to impose them is reasonable

File: lib/spack/spack/solver/concretize.lp (hunk 8 above)
Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp:524]: > This only says that if X provides a virtual Y, that Y must exist as a virtual.

I don't read it exactly like that, but rather: _if X satisfies the conditions to provide the virtual Y, then Y must exist as a virtual_. This is not always true. For instance, if in a solve `intel-parallel-studio` is used only for `mpi` there does not need to be a `lapack` virtual node in the answer set.

Comment (scheibelp) [reply lib/spack/spack/solver/concretize.lp:524]: > I don't read it exactly like that, but rather: if X satisfies the conditions to provide the virtual Y, then Y must exist as a virtual.

Agree.

> if in a solve intel-parallel-studio is used only for mpi there does not need to be a lapack virtual node in the answer set.

What do you think should happen for something like:

```
packages:
  lapack:
    require:
    - one_of: [openblas, netlib-lapack]

...

spack spec intel-parallel-studio+mkl
```

? 

(I don't think this is a counter example to your claim, but I am curious about the desired behavior).

Overall, it sounds like you'd prefer the second strategy I described in https://github.com/spack/spack/pull/39880#discussion_r1362499897 (namely, infer provider-specific requirements as provider-specific requirement groups).

Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp:524]: > What do you think should happen for something like: ...

I think it should install, since we asked for the `intel-parallel-studio+mkl` package explicitly.

**Comment by scheibelp** [comment]
This has been partially covered by https://github.com/spack/spack/pull/40639. Although it does more than that PR, I figure this should be closed until it can be connected to a wider issue.
